src/lib/linux-api/src/signal.rs

```diff
@@ -0,0 +1,1192 @@
+    let and = lhs & rhs;
+    assert!(and.has(Signal::SIGABRT));
+    assert!(!and.has(Signal::SIGSEGV));
+    assert!(!and.has(Signal::SIGALRM));
+}
+
+impl core::ops::BitAndAssign for SigSet {
+    fn bitand_assign(&mut self, rhs: Self) {
+        self.0 &= rhs.0
+    }
+}
+
+#[test]
+fn test_bitand_assign() {
+    let mut set = SigSet::from(Signal::SIGABRT) | SigSet::from(Signal::SIGSEGV);
+    set &= SigSet::from(Signal::SIGABRT) | SigSet::from(Signal::SIGALRM);
+    assert!(set.has(Signal::SIGABRT));
+    assert!(!set.has(Signal::SIGSEGV));
+    assert!(!set.has(Signal::SIGALRM));
+}
+
+impl core::ops::Not for SigSet {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        Self(!self.0)
+    }
+}
+
+#[test]
+fn test_not() {
+    let set = SigSet::from(Signal::SIGABRT) | SigSet::from(Signal::SIGSEGV);
+    let set = !set;
+    assert!(!set.has(Signal::SIGABRT));
+    assert!(!set.has(Signal::SIGSEGV));
+    assert!(set.has(Signal::SIGALRM));
+}
+
+pub type SignalHandlerFn = unsafe extern "C" fn(i32);
+pub type SignalActionFn = unsafe extern "C" fn(i32, *mut SigInfo, *mut core::ffi::c_void);
+
+pub enum SignalHandler {
+    Handler(SignalHandlerFn),
+    Action(SignalActionFn),
+    SigIgn,
+    SigDfl,
+}
+
+/// Expose for cbindgen APIs
+#[allow(non_camel_case_types)]
+pub type linux_sigaction = bindings::linux_sigaction;
+
+/// Compatible with kernel's definition of `struct sigaction`. Different from
+/// libc's in that `ksa_handler` and `ksa_sigaction` are explicitly in a union,
+/// and that `ksa_mask` is the kernel's mask size (64 bits) vs libc's larger one
+/// (~1000 bits for glibc).
+///
+/// We use the field prefix lsa_ to avoid conflicting with macros defined for
+/// the corresponding field names in glibc.
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct SigAction(linux_sigaction);
+unsafe impl VirtualAddressSpaceIndependent for SigAction {}
+unsafe impl Send for SigAction {}
+unsafe impl TransparentWrapper<linux_sigaction> for SigAction {}
+
+impl SigAction {
+    // Bindgen doesn't succesfully bind these constants; maybe because
+    // the macros defining them cast them to pointers.
+    //
+    // Copied from linux's include/uapi/asm-generic/signal-defs.h.
+    const SIG_DFL: usize = 0;
+    const SIG_IGN: usize = 1;
+
+    pub fn flags(&self) -> Option<SigActionFlags> {
+        SigActionFlags::from_bits(self.0.lsa_flags)
+    }
+
+    pub fn flags_retain(&self) -> SigActionFlags {
+        SigActionFlags::from_bits_retain(self.0.lsa_flags)
+    }
+
+    pub fn handler(&self) -> SignalHandler {
+        let as_usize = self.0.lsa_handler.map(|f| f as usize).unwrap_or(0);
+        if as_usize == Self::SIG_IGN {
+            SignalHandler::SigIgn
+        } else if as_usize == Self::SIG_DFL {
+            SignalHandler::SigDfl
+        } else if self.flags_retain().contains(SigActionFlags::SIGINFO) {
+            SignalHandler::Action(unsafe { core::mem::transmute(self.0.lsa_handler.unwrap()) })
```

Comment:
I think this transmute could be more explicit, maybe with the types defined explicitly. For example `transmute::<extern "C" fn(arg1: ::core::ffi::c_int), extern "C" fn(i32, *mut SigInfo, *mut core::ffi::c_void)>(lsa_handler.unwrap())`.

I also don't understand how this is correct since the arguments are completely different. I'm guessing this has something to do with the libc version using a union for `sa_handler` and `sa_sigaction`, but I think this needs some comments explaining what's going on and why it's safe.